functions/go/gatekeeper/validate.go

```diff
@@ -56,15 +58,16 @@ func createClient() (*opaclient.Client, error) {
 func gatherTemplates(objects []runtime.Object) ([]*templates.ConstraintTemplate, error) {
 	var templs []*templates.ConstraintTemplate
 	for _, obj := range objects {
-		ct, isConstraintTemplate := obj.(*v1beta1.ConstraintTemplate)
-		if !isConstraintTemplate {
-			continue
-		}
 		templ := &templates.ConstraintTemplate{}
-		if err := scheme.Convert(ct, templ, nil); err != nil {
-			return nil, err
+		switch obj.(type) {
+		case *v1.ConstraintTemplate, *v1beta1.ConstraintTemplate, *v1alpha1.ConstraintTemplate:
+			if err := scheme.Convert(obj, templ, nil); err != nil {
+				return nil, err
+			}
+			templs = append(templs, templ)
```

Comment:
I recommend substituting this implementation for [that of gktest](https://github.com/open-policy-agent/gatekeeper/blob/master/pkg/gktest/read_constraints.go#L112).  It uses the `ToVersionless()` func that's built into each version of ConstraintTemplate (v1alpha1, v1beta1, v1).

The upside of this is that `ToVersionless()` also [takes care of defaulting](https://github.com/open-policy-agent/frameworks/blob/master/constraint/pkg/apis/templates/v1/helpers.go#L11).  Switching CT schemas to be structural (which is required in k8s clusters >= v1.22) required us to do some special defaulting stuff under the hood.

Rather than one day deal with those problems, using `ToVersionless()` will save you the headache :)

Here's the logic you'll need, as well as I can write it in this comment without running it:

```golang
// outside of this function
type versionless interface {
	ToVersionless() (*templates.ConstraintTemplate, error)
}

// ...

// within the function
v, isVersionless := obj.(versionless)
if !isVersionless {
	continue
}

// this returns a &templates.ConstraintTemplate{}
template, err := v.ToVersionless()
if err != nil {
	// this means there's a bug probably, return to user
}
```

Reply:
Done!
Great suggestion!